Mathlib/NumberTheory/LSeries/Deriv.lean

```diff
@@ -0,0 +1,227 @@
+/-
+Copyright (c) 2024 Michael Stoll. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Michael Stoll
+-/
+import Mathlib.Analysis.Complex.LocallyUniformLimit
+import Mathlib.Data.Complex.ExponentialBounds
+import Mathlib.NumberTheory.LSeries.Convergence
+import Mathlib.Analysis.SpecialFunctions.Pow.Deriv
+
+/-!
+# Differentiability and derivatives of L-series
+
+## Main results
+
+* We show that the `LSeries` of `f` is differentiable at `s` when `re s` is greater than
+  the abscissa of absolute convergence of `f` (`LSeries.hasDerivAt`) and that its derivative
+  there is the negative of the `LSeries` of the point-wise product `log * f` (`LSeries.deriv`).
+
+* We prove similar results for iterated derivatives (`LSeries.iteratedDeriv`).
+
+* We use this to show that `LSeries f` is holomorphic on the right half-plane of
+  absolute convergence (`LSeries.analyticOn`).
+
+## Implementation notes
+
+We introduce `LSeries.logMul` as an abbreviation for the point-wise product `log * f`, to avoid
+the problem that this expression does not type-check. Similarly, we make use of the abbreviation
+`LSeries.logPowMul n f` for `log^n * f`.
+-/
+
+-- Should this go into its own file (because it needs `EReal`s, it doesn't fit well
+-- in some existing file), together with versions for the left, upper and lower half-planes?
+/-- An open right half-plane is open in the complex plane. -/
+lemma Complex.isOpen_rightHalfPlane (x : EReal) : IsOpen {z : ℂ | x < z.re} :=
+  isOpen_lt continuous_const <| EReal.continuous_coe_iff.mpr continuous_re
+
+open Complex LSeries
+
+/-!
+### Multiplication by (powers of) log
+-/
+
+/-- The (point-wise) product of `log : ℕ → ℂ` with `f`. -/
+noncomputable abbrev LSeries.logMul (f : ℕ → ℂ) (n : ℕ) : ℂ := log n * f n
+
+
+/-- The (point-wise) product of the `m`th power of `log` with `f`. -/
+noncomputable abbrev logPowMul (m : ℕ) (f : ℕ → ℂ) (n : ℕ) : ℂ :=
+  (log n) ^ m * f n
+
+lemma logPowMul_zero (f : ℕ → ℂ) : logPowMul 0 f = f := by
+  ext
+  simp only [logPowMul, pow_zero, one_mul]
+
+lemma logPowMul_succ (m : ℕ) (f : ℕ → ℂ) : logPowMul m.succ f = logMul (logPowMul m f) := by
+  ext
+  simp only [logPowMul, pow_succ, mul_assoc, logMul]
+
+lemma logPowMul_succ' (m : ℕ) (f : ℕ → ℂ) : logPowMul m.succ f = logPowMul m (logMul f) := by
+  ext m
+  simp only [logPowMul, pow_succ, logMul, ← mul_assoc]
+  rw [mul_comm (log m)]
```

Comment:
I suggest using the built-in API for iterating operations, `logMul^[n] f` (which is syntatic sugar for `Nat.iterate logMul n f`); this has its own lemmas `Function.iterate_zero`, `Function.iterate_succ` etc, so you don't need to roll your own.

Reply:
OK.